### PR TITLE
Allow Observables to return objects and still be differentiable

### PR DESCRIPTION
**Pull request by cvjjm**
Before making this into a proper RP I wanted to quickly get your opinion whether this change has a chance of making it into upstream.

Background: I realized that the only thing that is preventing users from constructing custom `Observables `that return an object (in my case an `OpenFermion.QubitOperator` is super handy) is a very small portion of code in `jacobian_tape.py` that makes assumptions about return types having certain properties.

With the rather minimal change below, I can have a device support a custom `Observable `that returns an `object `and and for which it supports the device gradient and use that in `return qml.expval(CustomObservable)` of a `QNode` and even differentiated that `QNode` with `qml.grad()`.

An alternative would be to implement `.dtype`, `__len__()`, and `.flatten()` in the class of the returned object and only change line 585 to take the `dtype` from g rather than a hardcoded `float`. Would that be preferred?

### CODE REVIEW

**Comment by josh146** [comment]
> With the rather minimal change below, I can have a device support a custom Observable that returns an object and and for which it supports the device gradient and use that in return qml.expval(CustomObservable) of a QNode and even differentiated that QNode with qml.grad().

@cvjjm this is great! I am in favour of this solution. The current code is written deliberately using duck typing, with the blocker here solely to distinguish between standard NumPy arrays and ragged arrays -- I'm not sure it makes sense to force objects to overwrite `flatten()` etc. just to keep to this.

My only concern is more one of maintenance. Best to leave a detailed comment here to make sure that future development adheres to this duck typing, and doesn't introduce any code that limits the returned object type.

**Comment by co9olguy** [comment]
Very much in favour of having a user's custom operations remain differentiable

File: pennylane/tape/jacobian_tape.py [hunk 1]
```diff
@@ -224,11 +224,18 @@ def _has_trainable_params(params, diff_methods):
     @staticmethod
     def _flatten_processing_result(g):
         """Flattens the output from processing_fn in parameter shift methods."""
-        if g.dtype is np.dtype("object"):
+        if hasattr(g, "dtype") and g.dtype is np.dtype("object"):
             # object arrays cannot be flattened; must hstack them
+            # check that g has attribute dtype first to allow for
+            # Observables that return other objects that do not have
+            # a dtype
```
Comment (josh146) [inline pennylane/tape/jacobian_tape.py:231]:
```suggestion
            # - Object arrays cannot be flattened; must hstack them.
            # - We also check that g has attribute dtype first to allow for
            #   Observables that return arbitrary objects.
```

File: pennylane/tape/jacobian_tape.py [hunk 2]
```diff
@@ -224,11 +224,18 @@ def _has_trainable_params(params, diff_methods):
...
             g = np.hstack(g)
 
-        return g.flatten()
+        if hasattr(g, "flatten"):
+            # flatten only if g supports flattening to allow for
+            # objects other than numpy ndarrays
+            return g.flatten()
+        return g
```
Comment (josh146) [inline pennylane/tape/jacobian_tape.py:238]: Not sure if this is too obfuscating, but we could avoid the if statement by doing this?
```python
getattr(g, "flatten", lambda: g)()
```

File: pennylane/tape/jacobian_tape.py [hunk 3]
```diff
@@ -580,9 +587,15 @@ def jacobian(self, device, params=None, **options):
...
-                jac = np.zeros((len(g), len(params)), dtype=float)
+                try:
+                    self._output_dim = len(g)
+                except TypeError:
+                    # if g has no len (e.g., because it is not a numpy.ndarray)
+                    # assume the dimension is 1
+                    self._output_dim = 1
+                # create the Jacobian matrix with appropriate dtype
+                dtype = float if isinstance(g, float) else np.object
+                jac = np.zeros((self._output_dim, len(params)), dtype=dtype)
```
Comment (josh146) [inline pennylane/tape/jacobian_tape.py:598]: @cvjjm I noticed that [quite a few of the tests are failing](https://github.com/PennyLaneAI/pennylane/pull/1291/checks?check_run_id=2566860157), specifically the `np.allclose` calls within the tests. I'm not 100% sure why, but I think the issue might be the logic here?

Should it instead be tied to the existence of a `__len__` attribute?
```python
dtype = float if hasattr(g, "__len__") else np.object
```

**Review by josh146 — CHANGES_REQUESTED** [review]
Thanks @cvjjm! This is a nice small modification that opens up a very nice feature. Before this can be merged in, we'll need to:

1. Work out why the tests are failing
2. Add a new test to ensure the new behaviour works (and catch it from breaking in the future).

File: pennylane/tape/jacobian_tape.py (hunk 2 above)
Comment (cvjjm) [reply pennylane/tape/jacobian_tape.py:237]: Too obfuscating for my taste :-)

File: pennylane/tape/jacobian_tape.py (hunk 3 above)
Comment (cvjjm) [reply pennylane/tape/jacobian_tape.py:597]: I wasn't sure what I need to put here to preserve the default behavior. I will look into this.

**Comment by cvjjm** [comment]
For me all the existing tests pass now and I have added a test of the newly supported behavior.

One last thing I am not totally happy with is that if I declare my QNode to accept an array as parameter like this:
```
        # force diff_method='parameter-shift' because otherwise
        # PennyLane swaps out dev for default.qubit.autograd
        @qml.qnode(dev, diff_method='parameter-shift')
        def qnode(x):
            qml.RY(x[0], wires=0)
            qml.RY(x[1], wires=0)
            return qml.expval(SpecialObservable(wires=0))

        @qml.qnode(dev, diff_method='parameter-shift')
        def reference_qnode(x):
            qml.RY(x[0], wires=0)
            qml.RY(x[1], wires=0)
            return qml.expval(qml.PauliZ(wires=0))

        params = np.array([0.2, 0.1])
        assert np.isclose(qnode(params).item().val, reference_qnode(params))
        assert np.isclose(qml.jacobian(qnode)(params).item().val, qml.jacobian(reference_qnode)(params))
```  
Differentiation does no longer work because of the following error:
```
Traceback (most recent call last):
  File "/.../pennylane/tests/tape/test_jacobian_tape.py", line 723, in test_special_observable_qnode_differentiation
    assert np.isclose(qml.jacobian(qnode)([0.2, 0.1]).item().val, qml.jacobian(reference_qnode)([0.2, 0.1]))
  File "/.../pennylane/pennylane/_grad.py", line 182, in _jacobian_function
    return _jacobian(func, argnum[0])(*args, **kwargs)
  File "/.../miniforge3/envs/.../lib/python3.8/site-packages/autograd/wrap_util.py", line 20, in nary_f
    return unary_operator(unary_f, x, *nary_op_args, **nary_op_kwargs)
  File "/.../miniforge3/envs/.../lib/python3.8/site-packages/autograd/differential_operators.py", line 59, in jacobian
    jacobian_shape = ans_vspace.shape + vspace(x).shape
TypeError: can only concatenate tuple (not "list") to tuple
```
I suspect there must be a way to teach autograd to construct proper VSpaces for my SpecialObject which would fix this, but I am unsure how. This however is not a problem of PennyLane. Just in case you have an idea how also this could be made to work it would be very nice if you told me how.

**Comment by josh146** [comment]
Thanks for the update @cvjjm!

> Just in case you have an idea how also this could be made to work it would be very nice if you told me how.

Unfortunately, while I come across this error many times, I am not 100% sure how to validate new type mappings for Autograd.

I can show you how I managed to add the `qml.numpy.tensor` class as a valid Autograd type though, perhaps this will be helpful:

```python
from autograd.tracer import Box
from autograd.numpy.numpy_boxes import ArrayBox
from autograd.numpy.numpy_vspaces import ComplexArrayVSpace, ArrayVSpace
from autograd.core import VSpace

def tensor_to_arraybox(x, *args):
    """Convert a :class:`~.tensor` to an Autograd ``ArrayBox``.

    Args:
        x (array_like): Any data structure in any form that can be converted to
            an array. This includes lists, lists of tuples, tuples, tuples of tuples,
            tuples of lists and ndarrays.

    Returns:
        autograd.numpy.numpy_boxes.ArrayBox: Autograd ArrayBox instance of the array
    """
    if isinstance(x, tensor):
        if x.requires_grad:
            return ArrayBox(x, *args)

        raise NonDifferentiableError(
            "{} is non-differentiable. Set the requires_grad attribute to True.".format(x)
        )

    return ArrayBox(x, *args)


Box.type_mappings[tensor] = tensor_to_arraybox
VSpace.mappings[tensor] = lambda x: ComplexArrayVSpace(x) if onp.iscomplexobj(x) else ArrayVSpace(x)
```

**Comment by cvjjm** [comment]
Great. I didn't know about the `Box.type_mappings` thing. I will play around with it, but for this PR it should be irrelevant. The last commit should have fixed the remaining code formatting problems. From my side this is now ready to merge.

**Comment by cvjjm** [comment]
Ups. Forgot the -l 100 flag when calling black... Now all should be well.

File: tests/tape/test_jacobian_tape.py [hunk 4]
```diff
@@ -633,6 +636,90 @@ def test_return_state_hessian_error(self):
...
+                    val = super().expval(qml.PauliZ(wires=0), **kwargs)
+                    return SpecialObject(val)
+
+                return super().expval(observable, **kwargs)
+
+        dev = DeviceSupporingSpecialObservable(wires=1, shots=None)
+
+        # force diff_method='parameter-shift' because otherwise
+        # PennyLane swaps out dev for default.qubit.autograd
+        @qml.qnode(dev, diff_method="parameter-shift")
```
Comment (josh146) [inline tests/tape/test_jacobian_tape.py:714]: Hopefully this will be fixed soon via #1334!

File: tests/tape/test_jacobian_tape.py [hunk 5]
```diff
@@ -633,6 +636,90 @@ def test_return_state_hessian_error(self):
...
+            qml.RY(x, wires=0)
+            return qml.expval(SpecialObservable(wires=0))
+
+        @qml.qnode(dev, diff_method="parameter-shift")
+        def reference_qnode(x):
+            qml.RY(x, wires=0)
+            return qml.expval(qml.PauliZ(wires=0))
+
+        assert np.isclose(qnode(0.2).item().val, reference_qnode(0.2))
+        assert np.isclose(qml.jacobian(qnode)(0.2).item().val, qml.jacobian(reference_qnode)(0.2))
```
Comment (josh146) [inline tests/tape/test_jacobian_tape.py:725]: A very thorough text example! This test will help ensure this (originally accidental) functionality remains working.

**Review by josh146 — APPROVED** [review]

**Comment by josh146** [comment]
@cvjjm, before merging, could I get you to:

- Add a short summary of the change to the changelog, along with a link back to the PR
- Add your name or GitHub username (whichever you prefer) to the Contributors section 

Thanks 🙂

**Pull request merged**